indexing-service/src/main/java/io/druid/indexing/overlord/DistributedTaskLockbox.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.indexing.overlord;
+
+import com.google.common.base.Optional;
+import com.google.inject.Inject;
+import io.druid.indexing.common.TaskLock;
+import io.druid.indexing.common.task.Task;
+import org.joda.time.Interval;
+
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Delegates all functionality to LocalTaskLockbox,
+ * makes additional check that lock is not acquired by RemoteTaskLockbox
+ */
+public class DistributedTaskLockbox implements TaskLockbox
+{
+  private static final int TASK_LOCKS_CACHE_SIZE = 1000;
```

Comment:
Forget to add this to config - should be configurable
Do not think that invalidation is critical, it's ok to have this cache with fix size.